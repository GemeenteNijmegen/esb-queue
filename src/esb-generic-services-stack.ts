import * as core from 'aws-cdk-lib';
import { CfnOutput } from 'aws-cdk-lib';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

export class esbGenericServicesStack extends core.Stack {

  constructor(scope: Construct, id: string, props: core.StackProps) {
    super(scope, id, props);
    core.Tags.of(this).add('cdkManaged', 'yes');
    core.Tags.of(this).add('Project', 'esb');

    /**
     * Custom KMS key for esb eform sqs connection.
     * Neccessary because the required permissions aren't included in the default key policy of the AWS managed KMS key for Amazon SQS, and you can't modify this policy.
     */
    const kmsKey = new kms.Key(this, 'esb-eform-sqs-key', {
      alias: 'esb/eform/sqs',
      enableKeyRotation: true,
      description: 'Custom KMS key for publishing messages from eform-submissions sns topic to the esb sqs queue.',
    });

    kmsKey.addToResourcePolicy(new iam.PolicyStatement({
      sid: 'Allow esb SQS Queue to receive messages from the eform-submissions SNS topic.',
      effect: iam.Effect.ALLOW,
      principals: [new iam.AnyPrincipal],
      actions: [
        'kms:Encrypt',
        'kms:Decrypt',
        'kms:GenerateDataKey',
      ],
      resources: ['*'],
    }));

    /**
     * Sns Topic from eform project: eform submissions sns topic.
     */
    const eformSubmissionsSnsTopic = sns.Topic.fromTopicArn(this, 'eform-submissions-sns-topic', ssm.StringParameter.valueForStringParameter(this, '/cdk/eform/SNSsubmissionsArn'));

    /**
     * Sqs Dead-Letter Queue: receives 'failed' messages to the esb eform submissions queue.
     */
    const eformSqsDlq = new sqs.Queue(this, 'esb-eform-submissions-dlq', {
      queueName: 'esb-eform-submissions-dlq',
      encryption: sqs.QueueEncryption.KMS,
      encryptionMasterKey: kmsKey,
    });

    /**
     * Sqs Queue: receives messages from eform submissions.
     */
    const eformSqs = new sqs.Queue(this, 'esb-eform-submissions-queue', {
      queueName: 'esb-eform-submissions-queue',
      encryption: sqs.QueueEncryption.KMS,
      encryptionMasterKey: kmsKey,
      deadLetterQueue: {
        queue: eformSqsDlq,
        maxReceiveCount: 1,
      },
    });

    new ssm.StringParameter(this, 'esb-eform-submissions-queue-arn', {
      parameterName: '/cdk/esb/queue/arn',
      stringValue: eformSqs.queueArn,
    });

    new CfnOutput(this, 'esb-eform-submissions-queue-arn-output', { // TODO remove this, however deployment fail right now as this output is deleted by the cloudformation upgrade.
      value: eformSqs.queueArn,
      exportName: 'esbAcceptance-esbGenericServices:esbAcceptanceesbGenericServicesExportsOutputFnGetAttesbeformsubmissionsqueue89F903F0ArnFBA3247A',
    });

    /**
     * Policy document: custom access policy for eform sqs.
     */
    const eformSqsPolicyDocument = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          actions: ['sqs:sendMessage'],
          effect: iam.Effect.ALLOW,
          conditions: {
            ArnEquals: {
              'aws:SourceArn': eformSubmissionsSnsTopic.topicArn,
            },
          },
          principals: [
            new iam.ServicePrincipal('sns.amazonaws.com'),
          ],
          resources: [
            eformSqs.queueArn,
          ],
        }),
      ],
    });
    new sqs.CfnQueuePolicy(this, 'eformSqsPolicyDocument', {
      queues: [eformSqs.queueUrl, eformSqsDlq.queueUrl],
      policyDocument: eformSqsPolicyDocument.toJSON(),
    });

    /**
     * Cloudwatch Alarm that triggers when message from eform submissions is delivered to the esb dlq.
     * Alarms are automatically picked up by aws-monitoring
     */
    new cloudwatch.Alarm(this, 'cloudwatch-alarm-esb-sqs-dlq', {
      metric: eformSqsDlq.metricNumberOfMessagesReceived({
        period: core.Duration.minutes(1), //TODO change to 30 sec because polling takes this amount of time?
      }),
      threshold: 0,
      evaluationPeriods: 1,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
      alarmName: 'esb-eform-sqs-dlq-alarm',
      alarmDescription: 'CloudWatch alarm that triggers when number of messages returned by calls to the ReceiveMessage action exceeds 0 on esb sqs dlq.',
    });

  }
}