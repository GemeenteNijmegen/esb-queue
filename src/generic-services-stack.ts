import * as sqs from '@aws-cdk/aws-sqs';
import * as core from '@aws-cdk/core';
// import * as kms from '@aws-cdk/aws-kms';

export class GenericServicesStack extends core.Stack {
  constructor(scope: core.Construct, id: string, props: core.StackProps) {
    super(scope, id, props);
    core.Tags.of(this).add('cdkManaged', 'yes');
    core.Tags.of(this).add('Project', 'esb');

    //TODO do we want this kms key? (to use in sqs as masterencryptionkey)
    //const kmsKey = kms.Key.fromKeyArn(this, 'eform-kms-key', ssm.StringParameter.valueForStringParameter(this, '/cdk/eform/eform/KMSKeysArn'));

    const eformSqsDlq = new sqs.Queue(this, 'esb-eform-submissions-dlq', {
      queueName: 'esb-eform-submissions-dlq',
      encryption: sqs.QueueEncryption.KMS,
    });

    new sqs.Queue(this, 'esb-eform-submissions-queue', {
      queueName: 'esb-eform-submissions-queue',
      encryption: sqs.QueueEncryption.KMS,
      deadLetterQueue: {
        queue: eformSqsDlq,
        maxReceiveCount: 1,
      },
    });

    //TODO add parameters
  }
}