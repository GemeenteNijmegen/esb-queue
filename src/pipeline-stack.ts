import * as codecommit from '@aws-cdk/aws-codecommit';
import * as core from '@aws-cdk/core';
import * as cdkpipelines from '@aws-cdk/pipelines';

class emptyStage extends core.Stage {

  constructor(scope: core.Construct, id: string, props: core.StageProps) {
    super(scope, id, props);

  }
}

class emptyyStage extends core.Stage {

  constructor(scope: core.Construct, id: string, props: core.StageProps) {
    super(scope, id, props);

  }
}

export class PipelineStack extends core.Stack {

  constructor(scope: core.Construct, id: string, props?: core.StackProps) {
    super (scope, id, props);
    core.Tags.of(this).add('cdkManaged', 'yes');
    core.Tags.of(this).add('project', 'esb');

    /**
         * Main repository
         */
    const repository = new codecommit.Repository(this, 'repository', {
      repositoryName: 'esb-repository',
    });

    if (process.env.BRANCH_NAME === 'acceptance') {
      const acceptancePipeline = new cdkpipelines.CodePipeline(this, 'acceptancePipeline', {
        pipelineName: 'esb-pipeline-acceptance',
        crossAccountKeys: true,
        synth: new cdkpipelines.ShellStep('Synth', {
          input: cdkpipelines.CodePipelineSource.codeCommit(repository, 'acceptance'),
          env: {
            BRANCH_NAME: '#{SourceVariables.BranchName}',
          },
          commands: [
            'yarn install --frozen-lockfile', //nodig om projen geinstalleerd te krijgen
            'npx projen build',
            'npx projen synth',
          ],
        }),
      });

      acceptancePipeline.addStage( new emptyStage(this, 'empty', {

      }),
      );

      acceptancePipeline.addStage( new emptyyStage(this, 'emptyy', {

      }),
      );
    }

    if (process.env.BRANCH_NAME === 'production') {
      const productionPipeline = new cdkpipelines.CodePipeline(this, 'productionPipeline', {
        pipelineName: 'esb-pipeline-production',
        crossAccountKeys: true,
        synth: new cdkpipelines.ShellStep('Synth', {
          input: cdkpipelines.CodePipelineSource.codeCommit(repository, 'production'),
          env: {
            BRANCH_NAME: '#{SourceVariables.BranchName}',
          },
          commands: [
            'yarn install --frozen-lockfile', //nodig om projen geinstalleerd te krijgen
            'npx projen build',
            'npx projen synth',
          ],
        }),
      });

      productionPipeline.addStage( new emptyStage(this, 'empty', {

      }),
      );

      productionPipeline.addStage( new emptyyStage(this, 'emptyy', {

      }),
      );
    }
  }
}