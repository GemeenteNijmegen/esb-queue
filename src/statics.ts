/**
 * Class with only static strings for naming of ssm paramaters
 * and re-use in stacks and constructs
 */
export class statics {
  //Statics for accountnames and numbers
  static readonly AWS_ACCOUNT_AUTH_ACCP: string = '315037222840';
  static readonly AWS_ACCCOUNT_AUTH_PROD: string = '196212984627';
  static readonly AWS_ACCOUNT_DEPLOYMENT: string = '418648875085';
  static readonly AWS_ACCOUNT_IAM: string = '098799052470';

  /**
   * Pipeline values
   */
  static readonly codeStarConnectionArn: string = 'arn:aws:codestar-connections:eu-west-1:418648875085:connection/4f647929-c982-4f30-94f4-24ff7dbf9766';
  static readonly projectRepo : string = 'GemeenteNijmegen/esb-queue';

  /**
   * SSM Parameters
   */
  static readonly ssmName_esbSqsArn: string = '/cdk/esb/queue/arn';
  static readonly ssmNmae_esfMailNotificationArn: string = '/cdk/esb/esf/notificationMail/arn';

}