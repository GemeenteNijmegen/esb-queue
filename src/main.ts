import * as core from '@aws-cdk/core';
import { PipelineStack } from './pipeline-stack';

const deploymentAccount = {
  account: '418648875085', // gemeentenijmegen-deployment
  region: 'eu-west-1',
};

const app = new core.App();

new PipelineStack(app, 'esb-pipeline', { env: deploymentAccount });

app.synth();