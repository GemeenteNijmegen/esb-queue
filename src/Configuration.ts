import { statics } from './statics';

/**
 * Custom Environment with obligatory accountId and region
 */
export interface Environment {
  account: string;
  region: string;
}

export interface Configurable {
  configuration: Configuration;
}

export interface Configuration {
  /**
     * The git branch name to which this configuration applies.
     */
  branchName: string;

  /**
     * Code star connection arn in the deployment environment
     */
  codeStarConnectionArn: string;

  /**
     * Deployment environment
     */
  deploymentEnvironment: Environment;

  /**
     * Target environment
     */
  targetEnvironment: Environment;

  /**
   * Domain name
   */
  domainName: string;

  /**
   * ESB stage name
   */
  esbStageName?: string;

  /**
   * Flag to indicate that the workload is
   * deployed to the new landingzone
   */
  isInNewLandingzone?: boolean;

}

export const configurations: { [key: string]: Configuration } = {
  'development-new-lz': {
    branchName: 'development-new-lz',
    codeStarConnectionArn: statics.gnBuildCodeStarConnectionArn,
    deploymentEnvironment: statics.gnBuildEnvironment,
    targetEnvironment: statics.gnWebformsDevEnvironment,
    domainName: 'dev.csp-nijmegen.nl',
    isInNewLandingzone: true,
  },
  'acceptance-new-lz': {
    branchName: 'acceptance-new-lz',
    codeStarConnectionArn: statics.gnBuildCodeStarConnectionArn,
    deploymentEnvironment: statics.gnBuildEnvironment,
    targetEnvironment: statics.gnWebformsAccpEnvironment,
    domainName: 'accp.csp-nijmegen.nl',
    isInNewLandingzone: true,
  },
  'main-new-lz': {
    branchName: 'main-new-lz',
    codeStarConnectionArn: statics.gnBuildCodeStarConnectionArn,
    deploymentEnvironment: statics.gnBuildEnvironment,
    targetEnvironment: statics.gnWebformsProdEnvironment,
    domainName: 'nijmegen.nl',
    isInNewLandingzone: true,
  },
  'development': {
    branchName: 'development',
    codeStarConnectionArn: statics.codeStarConnectionArn,
    deploymentEnvironment: statics.deploymentEnvironment,
    targetEnvironment: statics.sandboxEnvironment,
    domainName: 'dev.csp-nijmegen.nl',
  },
  'acceptance': {
    branchName: 'acceptance',
    codeStarConnectionArn: statics.codeStarConnectionArn,
    deploymentEnvironment: statics.deploymentEnvironment,
    targetEnvironment: statics.acceptanceEnvironment,
    domainName: 'accp.csp-nijmegen.nl',
    esbStageName: 'esbAcceptance',
  },
  'main': {
    branchName: 'main',
    codeStarConnectionArn: statics.codeStarConnectionArn,
    deploymentEnvironment: statics.deploymentEnvironment,
    targetEnvironment: statics.productionEnvironment,
    domainName: 'nijmegen.nl',
    esbStageName: 'esbProduction',
  },
};

export function getConfiguration(buildBranch: string) {
  const config = configurations[buildBranch];
  if (!config) {
    throw Error(`No configuration for branch ${buildBranch} found. Add a configuration in Configuration.ts`);
  }
  return config;
}