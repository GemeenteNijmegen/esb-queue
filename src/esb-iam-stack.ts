import * as core from 'aws-cdk-lib';
import { StackProps, aws_ssm as ssm } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Configurable } from './Configuration';
import { esbIAMMule } from './iam-structs/esb-mule-struct';
import { statics } from './statics';

export interface EsbIamStackProps extends StackProps, Configurable { }

export class EsbIamStack extends core.Stack {
  constructor(scope: Construct, id: string, props: EsbIamStackProps) {
    super(scope, id, props);

    const esbSqsArn = ssm.StringParameter.valueForStringParameter(this, statics.ssmName_esbSqsArn);
    const esfMailNotificationSqsArn = ssm.StringParameter.valueForStringParameter(this, statics.ssmName_esfMailNotificationArn);

    /**
     * IAM Construct for mule
     */
    new esbIAMMule(this, 'esbMule', {
      iamAccountPrincipal: statics.AWS_ACCOUNT_IAM,
      esbSqsArn: esbSqsArn,
      esfMailNotificationSqsArn: esfMailNotificationSqsArn,
      isInNewLandingzone: props.configuration.isInNewLandingzone,
    });
  }
}