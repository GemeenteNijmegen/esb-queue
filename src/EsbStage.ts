import { PermissionsBoundaryAspect } from '@gemeentenijmegen/aws-constructs';
import { Aspects, Stage, StageProps, Tags } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Configurable } from './Configuration';
import { esbGenericServicesStack } from './esb-generic-services-stack';
import { EsbIamStack } from './esb-iam-stack';
import { statics } from './statics';

export interface EsbStageProps extends StageProps, Configurable {}

export class EsbStage extends Stage {

  constructor(scope: Construct, id: string, props: EsbStageProps) {
    super(scope, id, props);
    Tags.of(this).add('cdkManaged', 'yes');
    Tags.of(this).add('project', statics.projectName);
    if (props.configuration.isInNewLandingzone) {
      Aspects.of(this).add(new PermissionsBoundaryAspect());
    }

    /**
     * Stack: esb generic services
     */
    const queues = new esbGenericServicesStack(this, 'esbGenericServices', {
      env: props.configuration.targetEnvironment,
      configuration: props.configuration,
    });

    /**
     * Stack: esb iam
     * Depends on GenericServicesStack
     */
    const iam = new EsbIamStack(this, 'esbIam', {
      env: props.configuration.targetEnvironment,
      configuration: props.configuration,
    });
    iam.addDependency(queues);

  }
}