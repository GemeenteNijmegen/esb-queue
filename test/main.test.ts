import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { PipelineStack } from '../src/pipeline-stack';

test('Snapshot', () => {
  const app = new App();
  const stack = new PipelineStack(app, 'test');

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});