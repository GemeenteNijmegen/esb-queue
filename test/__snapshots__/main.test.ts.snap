// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "repository9F1A3F0B": Object {
      "Properties": Object {
        "RepositoryName": "esb-repository",
        "Tags": Array [
          Object {
            "Key": "cdkManaged",
            "Value": "yes",
          },
          Object {
            "Key": "project",
            "Value": "esb",
          },
        ],
      },
      "Type": "AWS::CodeCommit::Repository",
    },
  },
}
`;
